### ISSUE TITLE
Activity calls from CreateRetryableClient are non-deterministic

### ISSUE THREAD

**Issue by cgillum** [Azure/durabletask#545]
## Issue
It appears that proxy objects generated by `OrchestrationContext.CreateRetriableClient` result in clients that invoke activities in a way that is non-deterministic. In particular, there appears to be threading problems with these clients that often result in stuck orchestrations or orchestrations that fail with errors like the following:

```
DurableTask.Core.Exceptions.NonDeterministicOrchestrationException: Non-Deterministic workflow detected: TaskScheduledEvent: 4 TaskScheduled Method5 
   at DurableTask.Core.TaskOrchestrationContext.HandleTaskScheduledEvent(TaskScheduledEvent scheduledEvent) in C:\GitHub\durabletask\src\DurableTask.Core\TaskOrchestrationContext.cs:line 268
   at DurableTask.Core.TaskOrchestrationExecutor.ProcessEvent(HistoryEvent historyEvent) in C:\GitHub\durabletask\src\DurableTask.Core\TaskOrchestrationExecutor.cs:line 141
   at DurableTask.Core.TaskOrchestrationExecutor.ExecuteCore(IEnumerable`1 eventHistory) in C:\GitHub\durabletask\src\DurableTask.Core\TaskOrchestrationExecutor.cs:line 82
   at DurableTask.Core.TaskOrchestrationContext.HandleTaskScheduledEvent(TaskScheduledEvent scheduledEvent) in C:\GitHub\durabletask\src\DurableTask.Core\TaskOrchestrationContext.cs:line 268
```

Previous maintainers of this repro have mentioned that this API is buggy.

The issue may be related to the use of [Dynamity](https://www.nuget.org/packages/Dynamitey) for dynamically invoking activity functions.

## Workarounds
* Use `OrchestrationContext.ScheduleWithRetry` - unfortunately this is not a type-safe way to call activity tasks.
* Use `OrchestrationContext.CreateClient` - unfortunately this doesn't support automatic retries.

**Comment by saguiitay** [comment]
Any update on this? Can you provide a snippet for using `OrchestrationContext.ScheduleWithRetry` (what name/version should I be using?)

**Comment by cgillum** [comment]
No updates, unfortunately, as we don't have the people resources to find a replacement for Dynamity. As far as `OrchestrationContext.ScheduleWithRetry()`, the name and version parameters correspond to the name of version that was inferred when you registered the activity using one of the overloads of [TaskHubWorker.AddTaskActivities()](https://github.com/Azure/durabletask/blob/e8b169ac97d2a2a4926412a1dcffbbca5e4c0e78/src/DurableTask.Core/TaskHubWorker.cs#L262-L307). If you just added them using the CLR type, then you can call them using the [ScheduleWithRetry](https://github.com/Azure/durabletask/blob/e8b169ac97d2a2a4926412a1dcffbbca5e4c0e78/src/DurableTask.Core/OrchestrationContext.cs#L135-L149) overload that takes a type parameter.

**Comment by saguiitay** [comment]
@cgillum any plans on handling this?

**Comment by cgillum** [comment]
No plans currently. Since none of our services depend on this API, a fix will likely need to come from the community. I wasn’t able to find a replacement for this after doing some quick research.

**Comment by aueloka** [comment]
@cgillum , I did a test with [Castle.Core](https://www.nuget.org/packages/Castle.Core/) library that seems a much better alternative to the current Dynamity. [Castle.Core](https://www.nuget.org/packages/Castle.Core/) is also used internally by the [Moq](https://github.com/moq/moq4) library.

Castle.Core supports creating proxy objects from interfaces, as well as abstract and non-sealed classes. With [Castle.Core](https://www.nuget.org/packages/Castle.Core/), this is worth the `RetryProxy.cs` would look like:

```c#
    internal class RetryProxy : IInterceptor
    {
        private readonly OrchestrationContext context;
        private readonly RetryOptions retryOptions;
        private readonly IDictionary<string, Type> returnTypes;

        /// <summary>
        /// Initializes a new instance of the <see cref="RetryProxy"/> class.
        /// </summary>
        /// <param name="context">The orchestration context.</param>
        /// <param name="retryOptions">The retry options.</param>
        public RetryProxy(OrchestrationContext context, RetryOptions retryOptions)
        {
            this.context = context;
            this.retryOptions = retryOptions;

            // If type can be determined by name
            this.returnTypes = typeof(T).GetMethods()
                .Where(method => !method.IsSpecialName)
                .GroupBy(method => method.Name)
                .Where(group => group.Select(method => method.ReturnType).Distinct().Count() == 1)
                .Select(group => new
                {
                    Name = group.Key,
                    ReturnType = group.Select(method => method.ReturnType).Distinct().Single(),
                })
                .ToDictionary(info => info.Name, info => info.ReturnType);
        }

        /// <inheritdoc/>
        public void Intercept(IInvocation invocation)
        {
            if (!this.returnTypes.TryGetValue(invocation.Method.Name, out Type? returnType))
            {
                throw new Exception("Method name '" + invocation.Method.Name + "' not known.");
            }

            if (!typeof(Task).IsAssignableFrom(returnType))
            {
                throw new InvalidOperationException($"Invoked method must return a task. Current return type is {invocation.Method.ReturnType}");
            }

            if (returnType == typeof(Task))
            {
                invocation.ReturnValue = this.InvokeWithRetry<object>(invocation);
                return;
            }

            returnType = invocation.Method.ReturnType.GetGenericArguments().Single();

            MethodInfo? invokeMethod = this.GetType().GetMethod("InvokeWithRetry", BindingFlags.Instance | BindingFlags.NonPublic);

            Debug.Assert(invokeMethod != null, "null");

            MethodInfo genericInvokeMethod = invokeMethod.MakeGenericMethod(returnType);
            invocation.ReturnValue = genericInvokeMethod.Invoke(this, new object?[] { invocation });

            return;
        }

        private async Task<TReturnType?> InvokeWithRetry<TReturnType>(IInvocation invocation)
        {
            Task<TReturnType> RetryCall()
            {
                return (Task<TReturnType>)invocation.GetConcreteMethod().Invoke(invocation.InvocationTarget, invocation.Arguments)!;
            }

            var retryInterceptor = new RetryInterceptor<TReturnType>(this.context, this.retryOptions, RetryCall);
            return await retryInterceptor.Invoke();
        }
    }
```

Similarly, the `ScheduleProxy.cs` would become:

```c#
internal sealed class ScheduleProxy : IInterceptor
    {
        private readonly OrchestrationContext context;
        private readonly Type interfaze;
        private readonly IDictionary<string, Type> returnTypes;
        private readonly bool useFullyQualifiedMethodNames;

        /// <summary>
        /// Initializes a new instance of the <see cref="ScheduleProxy"/> class.
        /// </summary>
        /// <param name="context">The orchestration context.</param>
        /// <param name="interface">The interface type.</param>
        public ScheduleProxy(OrchestrationContext context, Type @interface)
            : this(context, @interface, false)
        {
        }

        /// <summary>
        /// Initializes a new instance of the <see cref="ScheduleProxy"/> class.
        /// </summary>
        /// <param name="context">The orchestration context.</param>
        /// <param name="interface">The interface type.</param>
        /// <param name="useFullyQualifiedMethodNames">A flag indicating whether to use fully qualified method names.</param>
        public ScheduleProxy(OrchestrationContext context, Type @interface, bool useFullyQualifiedMethodNames)
        {
            this.context = context;
            this.interfaze = @interface;
            this.useFullyQualifiedMethodNames = useFullyQualifiedMethodNames;

            // If type can be determined by name
            this.returnTypes = this.interfaze.GetMethods()
                .Where(method => !method.IsSpecialName)
                .GroupBy(method => NameVersionHelper.GetDefaultName(method))
                .Where(group => group.Select(method => method.ReturnType).Distinct().Count() == 1)
                .Select(group => new
                {
                    Name = group.Key,
                    ReturnType = group.Select(method => method.ReturnType).Distinct().Single(),
                })
                .ToDictionary(info => info.Name, info => info.ReturnType);
        }

        /// <inheritdoc/>
        public void Intercept(IInvocation invocation)
        {
            if (!this.returnTypes.TryGetValue(invocation.Method.Name, out var returnType))
            {
                throw new Exception("Method name '" + invocation.Method.Name + "' not known.");
            }

            if (!typeof(Task).IsAssignableFrom(returnType))
            {
                throw new InvalidOperationException($"Invoked method must return a task. Current return type is {invocation.Method.ReturnType}");
            }

            Type[] genericArgumentValues = invocation.GenericArguments;
            List<object?> arguments = new(invocation.Arguments);

            foreach (var typeArg in genericArgumentValues)
            {
                arguments.Add(new TypeMetadata(typeArg.Assembly.FullName!, typeArg.FullName!));
            }

            object[] args = arguments.ToArray()!;

            string normalizedMethodName = NameVersionHelper.GetDefaultName(invocation.Method, this.useFullyQualifiedMethodNames);

            if (returnType == typeof(Task))
            {
                invocation.ReturnValue = this.context.ScheduleTask<object>(normalizedMethodName, NameVersionHelper.GetDefaultVersion(invocation.Method), args);
                return;
            }

            returnType = invocation.Method.ReturnType.GetGenericArguments().Single();

            MethodInfo scheduleMethod = typeof(OrchestrationContext).GetMethod(
                "ScheduleTask",
                new[] { typeof(string), typeof(string), typeof(object[]) }) ??
                throw new Exception($"Method 'ScheduleTask' not found. Type Name: {nameof(OrchestrationContext)}");

            MethodInfo genericScheduleMethod = scheduleMethod.MakeGenericMethod(returnType);

            var result = genericScheduleMethod.Invoke(this.context, new object[]
            {
                normalizedMethodName,
                NameVersionHelper.GetDefaultVersion(invocation.Method),
                args!,
            });

            invocation.ReturnValue = result;
            return;
        }
    }
```

With these changes, the `CreateClient<T>()` and `CreateRetryableClient<T>` would change like below:

```c#
public T CreateClient<T>(bool useFullyQualifiedMethodNames)
            where T : class
        {
            if (!typeof(T).IsInterface)
            {
                throw new InvalidOperationException("Pass in an interface.");
            }

            var proxy = new ProxyGenerator();

            IInterceptor interceptor = new ScheduleProxyV3(context, typeof(T), useFullyQualifiedMethodNames);
            return proxy.CreateInterfaceProxyWithoutTarget<T>(interceptor);
        }

public T CreateRetryableClient<T>(RetryOptions retryOptions, bool useFullyQualifiedMethodNames)
            where T : class
        {
            if (!typeof(T).IsInterface)
            {
                throw new InvalidOperationException("Pass in an interface.");
            }

            var proxyGen = new ProxyGenerator();

            IInterceptor scheduleProxy = new ScheduleProxyV3(context, typeof(T), useFullyQualifiedMethodNames);
            IInterceptor retryProxy = new RetryProxyV3(context, retryOptions);

            T scheduleInstance = proxyGen.CreateInterfaceProxyWithoutTarget<T>(scheduleProxy);
            return proxyGen.CreateInterfaceProxyWithTarget(scheduleInstance, retryProxy);
        }
```

**Comment by davidmrdavid** [comment]
@aueloka : Would you be interested in creating a PR for this?

**Comment by aueloka** [comment]
> @aueloka : Would you be interested in creating a PR for this?

Sure. I can. I have an open PR where I tried to introduce generic support related to this. I can modify that to include these changes if that's okay.

**Comment by davidmrdavid** [comment]
@aueloka - I think it would actually be better, review-wise, to keep them separate. Is that too much overhead? Perhaps we can work on that 2nd PR after the first one reaches a resolution

**Comment by aueloka** [comment]
> @aueloka - I think it would actually be better, review-wise, to keep them separate. Is that too much overhead? Perhaps we can work on that 2nd PR after the first one reaches a resolution

They both touch the same set of files. Essentially, whatever change I make relating to this issue would overwrite at lease 90 of the changes in the other PR.

**Comment by aueloka** [comment]
> > @aueloka - I think it would actually be better, review-wise, to keep them separate. Is that too much overhead? Perhaps we can work on that 2nd PR after the first one reaches a resolution
> 
> They both touch the same set of files. Essentially, whatever change I make relating to this issue would overwrite at lease 90 of the changes in the other PR.

Unless I proceed with this first and then update the first PR after this merges. I guess that could work.

**Comment by davidmrdavid** [comment]
Yeah I was thinking that doing them one after another would be better. I'll leave it to @cgillum, who I see is reviewing your PR, to make the final recommendation.

**Comment by aueloka** [comment]
@davidmrdavid / @cgillum , should this issue now be closed since the PR already went in? I didn't have permissions to include it in my PR or to close it.

**Comment by davidmrdavid** [comment]
I think so, though I'll leave it to @cgillum to make that call. He's currently out-of-office so it might be a few weeks until he gets to this.

**Comment by richardwhiuk** [comment]
@cgillum This should be closed - fixed in DTF 2.11 and later I think?

**Comment by cgillum** [comment]
@richardwhiuk yes, this was fixed in #545. Closing.
